File: golang/serializer/json_serializer.go
package serializer

import (
	"encoding/json"
	"fmt"

	"github.com/aws-glue-schema-registry/golang/client"
	"github.com/aws-glue-schema-registry/golang/model"
)

// JsonSerializer provides JSON serialization/deserialization
type JsonSerializer struct{}

// Serialize serializes a SalesforceAudit object to JSON format
func (s *JsonSerializer) Serialize(c *client.GlueSchemaRegistryClient, schemaName string, auditEvent *model.SalesforceAudit) ([]byte, error) {
	// Get schema definition from Glue Schema Registry
	schemaResponse, err := c.GetSchema(schemaName)
	if err != nil {
		return nil, fmt.Errorf("failed to get schema: %w", err)
	}

	latestVersion := *schemaResponse.LatestSchemaVersion
	_, err = c.GetSchemaVersion(schemaName, latestVersion)
	if err != nil {
		return nil, fmt.Errorf("failed to get schema version: %w", err)
	}

	// Note: In production, you might want to validate the JSON
	// against the schema definition before serialization using a JSON Schema validator

	// Serialize to JSON bytes
	jsonBytes, err := json.Marshal(auditEvent)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal JSON: %w", err)
	}

	return jsonBytes, nil
}

// Deserialize deserializes JSON data to a SalesforceAudit object
func (s *JsonSerializer) Deserialize(c *client.GlueSchemaRegistryClient, schemaName string, data []byte) (*model.SalesforceAudit, error) {
	// Get schema definition from Glue Schema Registry
	schemaResponse, err := c.GetSchema(schemaName)
	if err != nil {
		return nil, fmt.Errorf("failed to get schema: %w", err)
	}

	latestVersion := *schemaResponse.LatestSchemaVersion
	_, err = c.GetSchemaVersion(schemaName, latestVersion)
	if err != nil {
		return nil, fmt.Errorf("failed to get schema version: %w", err)
	}

	// Note: In production, you might want to validate the JSON
	// against the schema definition after deserialization using a JSON Schema validator

	// Deserialize from JSON bytes
	var auditEvent model.SalesforceAudit
	if err := json.Unmarshal(data, &auditEvent); err != nil {
		return nil, fmt.Errorf("failed to unmarshal JSON: %w", err)
	}

	return &auditEvent, nil
}

File: python/glue_schema_registry/json_serializer.py
"""Utility class for serializing and deserializing SalesforceAudit objects
using JSON schemas retrieved from Glue Schema Registry.
"""

import json
from typing import ByteString
from glue_schema_registry.client import GlueSchemaRegistryClient, SchemaRegistryException
from glue_schema_registry.model import SalesforceAudit


class JsonSerializer:
    """Utility class for JSON serialization/deserialization."""
    
    @staticmethod
    def serialize(client: GlueSchemaRegistryClient, schema_name: str, 
                  audit_event: SalesforceAudit) -> bytes:
        """Serializes a SalesforceAudit object to JSON format.
        
        Args:
            client: GlueSchemaRegistryClient instance
            schema_name: Name of the schema in the registry
            audit_event: SalesforceAudit object to serialize
            
        Returns:
            Serialized byte array (JSON)
            
        Raises:
            SchemaRegistryException: if serialization fails
        """
        try:
            # Get schema definition from Glue Schema Registry
            schema_response = client.get_schema(schema_name)
            latest_version = schema_response['LatestSchemaVersion']
            schema_version_response = client.get_schema_version(schema_name, latest_version)
            # Note: In production, you might want to validate against the schema
            
            # Serialize to JSON bytes
            json_str = json.dumps(audit_event.to_dict())
            return json_str.encode('utf-8')
            
        except Exception as e:
            raise SchemaRegistryException(f"Failed to serialize SalesforceAudit to JSON: {e}")
    
    @staticmethod
    def deserialize(client: GlueSchemaRegistryClient, schema_name: str, 
                   data: ByteString) -> SalesforceAudit:
        """Deserializes JSON data to a SalesforceAudit object.
        
        Args:
            client: GlueSchemaRegistryClient instance
            schema_name: Name of the schema in the registry
            data: Serialized JSON byte array
            
        Returns:
            Deserialized SalesforceAudit object
            
        Raises:
            SchemaRegistryException: if deserialization fails
        """
        try:
            # Get schema definition from Glue Schema Registry
            schema_response = client.get_schema(schema_name)
            latest_version = schema_response['LatestSchemaVersion']
            schema_version_response = client.get_schema_version(schema_name, latest_version)
            # Note: In production, you might want to validate against the schema
            
            # Deserialize from JSON bytes
            json_str = data.decode('utf-8')
            data_dict = json.loads(json_str)
            return SalesforceAudit.from_dict(data_dict)
            
        except Exception as e:
            raise SchemaRegistryException(f"Failed to deserialize JSON to SalesforceAudit: {e}")

